### DISCUSSION TITLE
Debugging User Defined Key Handlers

### DISCUSSION THREAD

**Discussion by TofuBug** [lzybkr/PSReadLine#3326] [Q&A]
I'm not sure if I'm just not putting together the right combination of Google search terms but I'm having a hard time figuring out how to actually Debug the ScriptBlock in a `Set-PSReadLineKeyHandler` cmdlet?

I'm adding key handlers with my profile, but I need to step through a few lines to better understand what's going on.

I've been running in VS Code and I can debug other PS scripts

**Comment by daxian-dbw** [comment]
@SeeminglyScience should be able to help here ;)
He made some improvements to VSCode recently which allows one to debug the key-binding script block in VSCode.

**Reply by SeeminglyScience** [reply]
Looks like a recent change may have broken it again, I have PowerShell/PowerShellEditorServices#1815 open to fix it again.

That said, when a preview hits with that PR, here's what you do (note 1-5 are one time set up per workspace only):

1. Open VSCode to the folder with your key handler in it
2. Add at least one break point inside your key handler
3. Open an external PowerShell window (ideally PowerShell 7.2+, but 5.1 will also work)
4. Press <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd> and start typing `Open 'launch.json'`
5. Add the following launch config (or click the button `Add Configuration...` and select `PowerShell: Attach to PowerShell Host Process` from the menu)
    ```json
        {
            "name": "PowerShell: Attach to PowerShell Host Process",
            "type": "PowerShell",
            "request": "attach",
            "runspaceId": 1
        },
    ```
6. On the left hand side menu of VSCode click the debug menu (play button with a little bug on it)
7. Select `PowerShell: Attach to PowerShell Host Process` from the menu at the top
    ![image](https://user-images.githubusercontent.com/24977523/170559855-6eb1145b-4a66-453a-9007-c16ea6a1937c.png)
8. Press <kbd>F5</kbd> and type the process ID of the external process you opened in step 3
9. In the external window from step 3, press the key associated with the key handler

If done correctly you will now be stopped on your breakpoint in the VSCode debugger:

![image](https://user-images.githubusercontent.com/24977523/170559690-a7f93d90-76bc-4164-81e7-26ae9f237ca4.png)

(PS: sorry for over explaining some parts, I know you're already using VSCode debugging so some of that is duplicate information)
